### PR TITLE
[HQ] Migration to ES 7.2.x

### PR DESCRIPTION
**Pull request by lwyszomi**
Hi @snopoke @emord 

I have done first changes to migrate CommCare to ElasticSearch 7.2.x, so currently this is Hold Of Merge but I would like to know that we go with correct direction?

Please let me know if you have any questions.

Regards,
Łukasz

### CODE REVIEW

File: corehq/ex-submodules/pillowtop/processors/elastic.py [hunk 1]
```diff
@@ -90,6 +90,8 @@ def send_to_elasticsearch(index, doc_type, doc_id, es_getter, name, data=None, r
 
     """
     data = data if data is not None else {}
+    del data['_id']
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/processors/elastic.py]: can you explain this change

File: corehq/ex-submodules/pillowtop/es_utils.py [hunk 2]
```diff
@@ -191,7 +187,12 @@ def initialize_mapping_if_necessary(es, index_info):
         pillow_logging.info("Initializing elasticsearch mapping for [%s]" % index_info.type)
         mapping = copy(index_info.mapping)
         mapping['_meta']['created'] = datetime.isoformat(datetime.utcnow())
-        mapping_res = es.indices.put_mapping(index_info.type, {index_info.type: mapping}, index=index_info.index)
+        mapping_res = es.indices.put_mapping(
+            index_info.type,
+            {index_info.type: mapping},
+            index=index_info.index,
+            include_type_name=True
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/es_utils.py]: what does this do?

**Review by snopoke — COMMENTED** [review]
I don't have a lot of context on the changes between the versions but it looks OK so far.

Apart from the inline questions:
* Is there going to be a separate PR for removing the Admin report code (I would think this could be done first to avoid having to 'upgrade' code / queries that aren't necessary).
* Do we have a rollout plan for this?
  * are these changes backwards compatible with the current version?
  * should we update the docker image?

File: corehq/ex-submodules/pillowtop/processors/elastic.py (hunk 1 above)
Comment (lwyszomi) [reply corehq/ex-submodules/pillowtop/processors/elastic.py]: @snopoke in the new version of ES field ```_id``` is reserved by ES and we can't send in the doc but only as a parameter

File: corehq/ex-submodules/pillowtop/es_utils.py (hunk 2 above)
Comment (lwyszomi) [reply corehq/ex-submodules/pillowtop/es_utils.py]: @snopoke more infomation why we need include_type_name you can find https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-put-mapping.html

**Comment by lwyszomi** [comment]
> I don't have a lot of context on the changes between the versions but it looks OK so far.
> 
> Apart from the inline questions:
> 
> * Is there going to be a separate PR for removing the Admin report code (I would think this could be done first to avoid having to 'upgrade' code / queries that aren't necessary).

Currently I'm focusing to prepare for my members in the team what we need to update to work with the new version of ES, and I wanted to confirm that we go with correct direction. I think this is a good suggestion to remove Admin Reports then fix queries in other places and the start working on the Admin Reports.

> * Do we have a rollout plan for this?
>   
>   * are these changes backwards compatible with the current version?
I didn't check this but probably these changes will not working with the current version
>   * should we update the docker image?
Currently I'm using image prepared by ES team:
https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html

File: requirements/requirements.txt [hunk 3]
```diff
@@ -62,7 +62,7 @@ djangorestframework==3.5.4
 dnspython==1.15.0
 docutils==0.14            # via botocore
 dropbox==9.3.0
-elasticsearch==1.9.0
+elasticsearch==6.4.0
```
Comment (proteusvacuum) [inline requirements/requirements.txt]: this version doesn't match the PR description. Are we upgrading directly to 7.2 or 6.4?

File: corehq/pillows/mappings/xform_mapping.py [hunk 4]
```diff
@@ -15,32 +15,32 @@
         "created": '2013-08-13',
     },
     "properties": {
-        'doc_type': {'type': 'string'},
+        'doc_type': {'type': 'text'},
```
Comment (proteusvacuum) [inline corehq/pillows/mappings/xform_mapping.py]: it would be great to convert all of these mappings to JSON files as we have for app and case mappings.

**Review by proteusvacuum — COMMENTED** [review]
This looks like a good start to me, although I'm curious how this is going to be rolled out if this code isn't backwards compatible with the old version. When we'd talked about this with Preethi we were going to do something like run two ES clusters at once and slowly switch over where it was reading from once all the data was there. Not sure if that is still the plan here. 

FYI @biyeun

File: corehq/pillows/mappings/xform_mapping.py (hunk 4 above)
Comment (lwyszomi) [reply corehq/pillows/mappings/xform_mapping.py]: I had this in my mind but I focused to fix errors in queries and mappings.

File: requirements/requirements.txt (hunk 3 above)
Comment (lwyszomi) [reply requirements/requirements.txt]: I will change this, documentation on elasticsearch-py.readthedocs.io is not updated and I didn't notice that we have 7.0.2 version of this library on pypi.org

**Comment by lwyszomi** [comment]
> This looks like a good start to me, although I'm curious how this is going to be rolled out if this code isn't backwards compatible with the old version. When we'd talked about this with Preethi we were going to do something like run two ES clusters at once and slowly switch over where it was reading from once all the data was there. Not sure if that is still the plan here.
> 
> FYI @biyeun

I didn't know what is the plan, I got information that we should update elasticsearch to the newest stable version to refactor Admin Reports.

File: corehq/apps/es/es_query.py [hunk 5]
```diff
@@ -167,9 +167,9 @@ def __init__(self, index=None, debug_host=None, es_instance_alias=ES_DEFAULT_INS
         self._aggregations = []
         self.es_instance_alias = es_instance_alias
         self.es_query = {"query": {
-            "filtered": {
-                "filter": {"and": []},
-                "query": queries.match_all()
+            "bool": {
+                "filter": queries.match_all(),
+                "must": []
```
Comment (esoergel) [inline corehq/apps/es/es_query.py]: This change also breaks `_query`, `set_query`, and `get_query` in this class, which expect the old format.

File: corehq/apps/es/es_query.py [hunk 6]
```diff
@@ -249,7 +249,7 @@ def scroll(self):
 
     @property
     def _filters(self):
-        return self.es_query['query']['filtered']['filter']['and']
+        return self.es_query['query']['bool']['must']
```
Comment (esoergel) [inline corehq/apps/es/es_query.py]: Shouldn't the filters go in the `filter` block of the `bool` query, not `must`, since they aren't used for scoring?

**Review by esoergel — COMMENTED** [review]
@lwyszomi you wrote:
> I got information that we should update elasticsearch to the newest stable version to refactor Admin Reports.

Do you have a ticket for that (or for this work overall)?  Admin reports are on the way out, and I believe are ready to be fully removed, which should make this process a lot easier.  The code I'm more concerned with is [the API's use of elasticsearch](https://github.com/dimagi/commcare-hq/blob/19e35a6e6243ef471acfc309515e3c6126acf532/corehq/apps/api/es.py).  That doesn't use the `ESQuery` wrapper, so it'll need to either be converted to `ESQuery` or manually upgraded.

File: corehq/apps/es/es_query.py (hunk 6 above)
Comment (lwyszomi) [reply corehq/apps/es/es_query.py]: @esoergel If I good understand how the ```bool``` query works in new version then ```bool must``` is correct and should work the same as we have previously. In ES 6.4 ```filtered```, ```and``` and ```or``` filters were deprecated and was replaced by bool query. https://www.elastic.co/guide/en/elasticsearch/reference/6.4/query-dsl-bool-query.html

File: corehq/apps/es/es_query.py (hunk 5 above)
Comment (lwyszomi) [reply corehq/apps/es/es_query.py]: @esoergel I know that these changes break a lot of in application, I prepared this PR to make sure that we go to correct direction which Dimagi want to go.

**Comment by lwyszomi** [comment]
> Do you have a ticket for that (or for this work overall)? Admin reports are on the way out, and I believe are ready to be fully removed, which should make this process a lot easier. The code I'm more concerned with is [the API's use of elasticsearch](https://github.com/dimagi/commcare-hq/blob/19e35a6e6243ef471acfc309515e3c6126acf532/corehq/apps/api/es.py). That doesn't use the `ESQuery` wrapper, so it'll need to either be converted to `ESQuery` or manually upgraded.

I talked about this work with Ashley at my weekly meetings a couple of months ago, but always we had more important cases to do. Now I want to prepare tasks for my new members in SolDevelo team for future work and I want to show the direction which they should go to, after my leaving from the team.

So all the suggestions which you have are very important.

FYI @dborowiecki @MaciejChoromanski

File: corehq/apps/es/es_query.py (hunk 6 above)
Comment (esoergel) [reply corehq/apps/es/es_query.py]: From that docs page:
> **must** The clause (query) must appear in matching documents and will contribute to the score.
> **filter** The clause (query) must appear in matching documents. However unlike must the score of the query will be ignored. Filter clauses are executed in filter context, meaning that scoring is ignored and clauses are considered for caching.

I take that to mean both will work, in that they'll return the same results, but the difference is that `must` will contribute to scoring, and those clauses won't be cached.  The `filter` context is used for filtering structured data.  Here's [the docs page on the difference](https://www.elastic.co/guide/en/elasticsearch/reference/6.4/query-filter-context.html).

The [docs for migrating](https://www.elastic.co/guide/en/elasticsearch/reference/6.4/query-dsl-filtered-query.html) from `filtered` to `bool` also say to move filters to the `filter` block.

**Pull request closed without merging**